5_Stable_implementation/flip.cpp

```diff
@@ -0,0 +1,332 @@
+}
+
+static torch::TensorIterator make_index_iterator(const AdvancedIndex& info) {
+  torch::TensorIteratorConfig config;
+  config.set_check_mem_overlap(false)
+        .check_all_same_dtype(false)
+        .declare_static_dtype_and_device(info.src.scalar_type(), info.src.device())
+        .add_output(torch::Tensor())
+        .add_input(info.src);
+  for (auto& index : info.indices) {
+    config.add_input(index);
+  }
+  return config.build();
+}
+
+void index_kernel(torch::TensorIterator& iter, torch::IntArrayRef index_size,
+                  torch::IntArrayRef index_stride, bool serial_execution=false) {
+    using scalar_t = uint8_t;
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(torch::ScalarType::Half, torch::ScalarType::Bool, torch::ScalarType::BFloat16,
+      iter.dtype(), "flip_cpu", [&] {
+        int ntensor = iter.ntensors();
+        // When launch the index parallel version, set a relative samll grain size less than the INTERNAL::GRAIN_SIZE
+        // to make the whole available thread numbers get more balanced work load and a better cache location.
+        // The grain size here is chosen by the op benchmark to overcome the thread launch overhead
+        const int index_parallel_grain_size = 3000;
+        auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+            auto indexer = Indexer(ntensor - 2, &data[2], &strides[2], index_size, index_stride);
+            char* dst = data[0];
+            char* src = data[1];
+
+            // int64_t offset = indexer.get(0);
+
+            for (int64_t i = 0; i < n; i++) {
+                int64_t offset = indexer.get(i);
+                *(scalar_t*)(dst + strides[0] * i) = *(scalar_t*)(src + strides[1] * i + offset);
+            }
+        };
+
+        if (serial_execution) {
+            iter.serial_for_each(loop, {0, iter.numel()});
+        } else {
+            iter.for_each(loop, index_parallel_grain_size);
+        }
+      });
+}
+
+/**
+ * Flip a N-dimensional tensor around a given set of dimensions.
+ * \param input the input tensor to flip
+ * \param flip_dims a list of dimensions used to flip the tensor.
+ **/
+torch::Tensor generalized_flip(torch::Tensor input, torch::IntArrayRef flip_dims) {
+    // std::sort(flip_dims.begin(), flip_dims.end());
+    std::vector<int64_t> dims;
+    for(int64_t dim: flip_dims) {
+        dims.push_back(dim);
+    }
+
+    std::sort(dims.begin(), dims.end());
+
+    std::vector<torch::Tensor> indices;
+    std::vector<int64_t> transposed_indices;
+    // std::vector<int64_t> sizes;
+    std::tie(input, indices, transposed_indices) = build_indices(input, dims);
+
+    auto advanced_index = AdvancedIndex(input, indices);
+    auto iter = make_index_iterator(advanced_index);
+    index_kernel(iter, advanced_index.indexed_sizes, advanced_index.indexed_strides);
+    auto result = iter.output();
+
+    if(transposed_indices.size() > 0) {
+      result = result.permute(transposed_indices);
+    }
```

Comment:
This can probably be removed if we remove the `transpose_to_front` that we have in the beginning of the implementation